sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -242,18 +243,19 @@ public void splitRestriction(
     }
 
     @NewTracker
-    public RestrictionTracker<BoundedSource<T>, Object[]> restrictionTracker(
+    public RestrictionTracker<BoundedSource<T>, TimestampedValue<T>[]> restrictionTracker(
         @Restriction BoundedSource<T> restriction, PipelineOptions pipelineOptions) {
       return new BoundedSourceAsSDFRestrictionTracker<>(restriction, pipelineOptions);
     }
 
     @ProcessElement
     public void processElement(
-        RestrictionTracker<BoundedSource<T>, Object[]> tracker, OutputReceiver<T> receiver)
+        RestrictionTracker<BoundedSource<T>, TimestampedValue<T>[]> tracker,
+        OutputReceiver<T> receiver)
         throws IOException {
-      Object[] out = new Object[1];
+      TimestampedValue<T>[] out = new TimestampedValue[1];
       while (tracker.tryClaim(out)) {
-        receiver.output((T) out[0]);
+        receiver.outputWithTimestamp(out[0].getValue(), out[0].getTimestamp());
```

Comment:
Is the timestamp for the output timestamp of records from BoundedSource?

Reply:
yes